회원정보 수정, 비밀빈호 찾기 기능

kingseungil opened the pull request:
#️⃣ 연관된 이슈
#10

📝 작업 내용
- 회원 정보 수정
- 비밀번호 찾기
	- 이메일로 임시 비밀번호 발송
	- 발송된 임시 비밀번호로 로그인 후 비밀번호 변경 api 호출
- 비밀번호 변경

💬 리뷰 요구사항

kingseungil committed:
fix: Redis에서 accessToken 업데이트 로직 제거

- 새로운 accessToken이 발급될 때마다 Redis에서 accessToken을 업데이트하는 로직 제거(임시)
- 우선 백엔드만 구현하기때문에 편의성을 위해 이렇게 바꿈. 나중에 클라이언트 쪽에서 새로운 엑세스토큰을 response header에서 가져와서 request 보낼 때 담아주는 작업 필요할듯

kingseungil committed:
chore: entity클래스에 @Table 추가

kingseungil committed:
feat: 회원 정보 수정 api (#10)

- 프로필이미지 수정 시 s3 delete 추가

kingseungil committed:
test: 회원 정보 수정 api (#10)

- DB cleaner 추가 (테스트 후 데이터일괄 삭제)
- 공통 로직 BaseController로 분리

kingseungil committed:
feat: 임시 비밀번호 메일 api (#10)

kingseungil committed:
feat: 비밀번호 재설정 (#10)

kingseungil committed:
test: 비밀번호 재설정 (#10)

kingseungil committed:
test: 비밀번호 찾기 (#10)

eTaphee commented on src/main/java/matgo/member/application/MemberService.java:
```diff
@@ -99,4 +104,61 @@ private void saveEmailVerification(String verificationCode, Member member) {
+        }
+
+        if (!profileImage.isEmpty()) {
+            s3Service.delete(member.getProfileImage());
```
요부분에서 궁금한게 있습니다!

가능성은 적겠지만 트랜잭션이 실패하게 되면, s3 이미지가 교체가 될거 같은데 맞을까요??

`updateProfileImageIfPresent` 에서 이미지는 지워젔으나, 업로드가 실패한경우가 있을 수도 있을거같아요.

kingseungil replied:
@eTaphee 님
좋은 의견 감사합니다! 그 부분은 생각도 못하고있었네요..! 
실제로 말씀해주신 오류가 발생했습니다!
그래서 일단  삭제 -> 업로드 로직을 업로드 -> 삭제로 순서만 변경해줬네요! 일단은 가장 간단하게 해결해봤습니당

eTaphee commented on src/main/java/matgo/member/presentation/MemberController.java:
```diff
@@ -38,13 +41,24 @@ public ResponseEntity<String> register(
+    public ResponseEntity<Void> updateMember(
+      @AuthenticationPrincipal UserDetails userDetails,
+      @Valid @RequestPart MemberUpdateRequest memberUpdateRequest,
+      @RequestPart(required = false) MultipartFile profileImage
```
프로필 이미지에 확장자랑 크기에 대한 유효성 검사가 필요하지 않을까요?!

kingseungil replied:
@eTaphee 님
앗 이부분도 생각못하고 있었네요!
덕분에 이미지 유효성 검사 로직도 추가했습니다! 감사합니다 :)

eTaphee commented on src/main/resources/static/docs/matgo-api.yaml:
이 파일은 swagger에서 사용되는 파일같은데, summary, description, 응답 정보, 예시는 어디서 가져오나요??

주석을 찾아봤는데, 안보여서요 여쭤봐요.

eTaphee replied:
MemberDocument 클래스에 있었네요! 자문자답..

kingseungil committed:
fix: 회원정보 수정 시 이미지 업로드 부분 (#10)

- 프로필사진 업데이트 시, s3 image 삭제 후 새로운 이미지 s3에 업로드를 해주고 있었는데, 업로드에서 에러가 나면 s3 파일만 삭제되고 디비에는 반영이 안되는 문제

kingseungil committed:
feat: 이미지 유효성 검사 (#10)

- 확장자 검사
- 이미지 크기 제한(5MB)

kingseungil committed:
fix: 테스트환경에서 mail 안보내지도록 수정

mingming-mentor commented on src/main/java/matgo/global/filesystem/AllowedFileExtension.java:
```diff
@@ -0,0 +1,14 @@
+        String upperCaseExtension = extension.toUpperCase();
+
+        return Arrays.stream(AllowedFileExtension.values())
+                     .anyMatch(allowedFileExtension -> allowedFileExtension.name().equals(upperCaseExtension));
```
enum의 경우 == 로 비교해주어도 됩니다.

mingming-mentor commented on src/main/java/matgo/member/application/MemberService.java:
```diff
@@ -99,4 +104,64 @@ private void saveEmailVerification(String verificationCode, Member member) {
+    }
+
+    private void updateProfileImageIfPresent(Member member, MultipartFile profileImage) {
+        if (profileImage == null) {
+            return;
+        }
+
+        if (!profileImage.isEmpty()) {
+            String newImageURL = uploadAndGetImageURL(profileImage);
+            String oldImageURL = member.getProfileImage();
+            member.changeProfileImage(newImageURL);
+            s3Service.delete(oldImageURL);
+        }
```
합쳐서 `if(profileImage != null && !profileImage.isEmpty())` 로 체크해도 좋을듯 하네요

mingming-mentor reviewed (approved):
수고 많으셨습니다!

kingseungil committed:
chore: 피드백 반영 (#17)

- 가독성 향상
